File: VB/WindowsApplication1/UploadHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.IO
Imports System.Diagnostics

Namespace WindowsApplication1
	Public Class UploadHelper
		Public Event CurrentFileChanged As EventHandler

		Private _DataSource As DataTable
		Public Property DataSource() As DataTable
			Get
				Return _DataSource
			End Get
			Set(ByVal value As DataTable)
				_DataSource = value
			End Set
		End Property


		Public Sub New(ByVal dataSource As DataTable)
			_DataSource = dataSource
		End Sub

		Private _CurrentFile As UploadedFile
		Public Property CurrentFile() As UploadedFile
			Get
				Return _CurrentFile
			End Get
			Set(ByVal value As UploadedFile)
				_CurrentFile = value
				OnCurrentFileChanged()
			End Set
		End Property

		Private Sub OnCurrentFileChanged()
			RaiseEvent CurrentFileChanged(Me, EventArgs.Empty)
		End Sub

		Public Sub UploadFile()
			UploadFile(-1)
		End Sub
		Public Sub UploadFile(ByVal dataSourceRowIndex As Integer)
			If CurrentFile IsNot Nothing Then
				Dim values() As Object = { CurrentFile.FileName, CurrentFile.FileSize, CurrentFile.FileContent }
				If dataSourceRowIndex = -1 Then
					DataSource.Rows.Add(values)
				Else
					DataSource.Rows(dataSourceRowIndex).ItemArray = values
				End If
			End If
			CurrentFile = Nothing
		End Sub

		Public Sub ChooseFile()
			If CurrentFile IsNot Nothing Then
				MessageBox.Show(String.Format("Please upload the{0} file", CurrentFile.FileName))
				Return
			End If
			Dim fd As New OpenFileDialog()
			If fd.ShowDialog() = DialogResult.OK Then
				CurrentFile = New UploadedFile(fd.FileName)
			End If

		End Sub

		Public Sub PostNewFile(ByVal dataSourceRowIndex As Integer)
			ChooseFile()
			UploadFile(dataSourceRowIndex)
		End Sub
	End Class
End Namespace

File: CS/WindowsApplication1/UploadHelper.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.IO;
using System.Diagnostics;

namespace WindowsApplication1
{
    public class UploadHelper
    {
        public event EventHandler CurrentFileChanged;

        private DataTable _DataSource;
        public DataTable DataSource
        {
            get { return _DataSource; }
            set { _DataSource = value; }
        }


        public UploadHelper(DataTable dataSource)
        {
            _DataSource = dataSource;
        }

        private UploadedFile _CurrentFile;
        public UploadedFile CurrentFile
        {
            get { return _CurrentFile; }
            set { _CurrentFile = value; OnCurrentFileChanged(); }
        }

        private void OnCurrentFileChanged()
        {
            if (CurrentFileChanged != null)
                CurrentFileChanged(this, EventArgs.Empty);
        }

        public void UploadFile()
        {
            UploadFile(-1);
        }
        public void UploadFile(int dataSourceRowIndex)
        {
            if (CurrentFile != null)
            {
                object[] values = new object[] { CurrentFile.FileName, CurrentFile.FileSize, CurrentFile.FileContent };
                if (dataSourceRowIndex < -1)
                    DataSource.Rows.Add(values);
                else
                    DataSource.Rows[dataSourceRowIndex].ItemArray = values;
            }
            CurrentFile = null;
        }

        public void ChooseFile()
        {
            if (CurrentFile != null)
            {
                MessageBox.Show(String.Format("Please upload the{0} file", CurrentFile.FileName));
                return;
            }
            OpenFileDialog fd = new OpenFileDialog();
            if (fd.ShowDialog() == DialogResult.OK)
            {
                CurrentFile = new UploadedFile(fd.FileName);
            }

        }

        public void PostNewFile(int dataSourceRowIndex)
        {
            ChooseFile();
            UploadFile(dataSourceRowIndex);
        }
    }
}
